training/deepspeech_training/evaluate.py

```diff
@@ -43,47 +43,44 @@ def sparse_tuple_to_texts(sp_tuple, alphabet):
     return [alphabet.Decode(res) for res in results]
 
 
-def evaluate(test_csvs, create_model):
+def evaluate(test_csvs, model_ctor):
     if FLAGS.scorer_path:
         scorer = Scorer(FLAGS.lm_alpha, FLAGS.lm_beta,
                         FLAGS.scorer_path, Config.alphabet)
     else:
         scorer = None
 
-    test_sets = [create_dataset([csv],
-                                batch_size=FLAGS.test_batch_size,
-                                train_phase=False,
-                                reverse=FLAGS.reverse_test,
-                                limit=FLAGS.limit_test) for csv in test_csvs]
-    iterator = tfv1.data.Iterator.from_structure(tfv1.data.get_output_types(test_sets[0]),
-                                                 tfv1.data.get_output_shapes(test_sets[0]),
-                                                 output_classes=tfv1.data.get_output_classes(test_sets[0]))
-    test_init_ops = [iterator.make_initializer(test_set) for test_set in test_sets]
+    with tfv1.Session(config=Config.session_config) as session:
+        test_sets = [create_dataset([csv],
+                                    batch_size=FLAGS.test_batch_size,
+                                    train_phase=False,
+                                    reverse=FLAGS.reverse_test,
+                                    limit=FLAGS.limit_test) for csv in test_csvs]
+        iterator = tfv1.data.Iterator.from_structure(tfv1.data.get_output_types(test_sets[0]),
+                                                     tfv1.data.get_output_shapes(test_sets[0]),
+                                                     output_classes=tfv1.data.get_output_classes(test_sets[0]))
+        test_init_ops = [iterator.make_initializer(test_set) for test_set in test_sets]
 
-    batch_wav_filename, (batch_x, batch_x_len), batch_y = iterator.get_next()
+        batch_wav_filename, (batch_x, batch_x_len), batch_y = iterator.get_next()
 
-    # One rate per layer
-    no_dropout = [None] * 6
-    logits, _ = create_model(batch_x=batch_x,
-                             seq_length=batch_x_len,
-                             dropout=no_dropout)
+        model = model_ctor()
+        logits, _ = model(batch_x)
 
-    # Transpose to batch major and apply softmax for decoder
-    transposed = tf.nn.softmax(tf.transpose(a=logits, perm=[1, 0, 2]))
+        # Apply softmax and transpose to batch major for batch decoder
+        transposed = tf.transpose(tf.nn.softmax(logits), [1, 0, 2])
```

Comment:
we moved from `softmax(transpose)` to `transpose(softmax)`, is it on purpose?

Reply:
Not on purpose, just a result of iterating on the patch. But it's equivalent, softmax is applied to the last dimension which is not (and was not) transposed here.

Reply:
right, but I guess for ease of clarity keep it as besore if better?

Reply:
Yeah, will do.